### PR TITLE
Add evented colormap

### LINKED ISSUE #862: Generic colormap representation
## 🚀 Feature
Right now in our model files we use the vispy colormap object. It would be nice to completely separate our model files from vispy as noted here https://github.com/napari/napari/pull/686#discussion_r363550535 in discussions around serializing the viewer and round-trips.

After #782 (which does a similar refactor but for colors) and #854 (which adds a camera model) both get merged our models will be vispy independent.

Note I'm not sure yet what this representation should be or if vispy colormaps are serializable as they are now, but am making this issue to discuss this.

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This is a super preliminary attempt at adding our own named colormap object that is pure python (no shader code) and vispy independent. Will close #862, can up in context of https://github.com/napari/napari/pull/1475#issuecomment-665295798.

I've got a bunch of stuff still to do, so not ready for review yet, but just putting out there in case @tlambert03 wants to see. Note this includes #1475, but definitely didn't need to and all sorts of stuff isn't hooked up yet!!

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
I can’t seem to find the main color map object... is this missing /utils/colormaps/colormap.py?

**Comment by sofroniewn** [comment]
> I can’t seem to find the main color map object... is this missing /utils/colormaps/colormap.py?

I forgot to do git add .... done now

File: napari/utils/colormaps/colormap.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        Colormap interpolation mode, either `linear` or
+        `zero`.
+    """
+
+    name: str = 'undefined'
+    colors: Property[np.ndarray, None, transform_color] = np.zeros((2, 4))
+    controls: Property[np.ndarray, None, np.asarray] = np.asarray(
+        [0, 1]
+    )  # Not yet implemented
+    interpolation: str = 'linear'  # Only linear supported
```
Comment (tlambert03) [inline napari/utils/colormaps/colormap.py]: huh... that is quite nice :joy:

File: napari/utils/colormaps/colormaps.py [hunk 2]
```diff
@@ -438,4 +422,35 @@ def ensure_colormap_tuple(colormap: ValidColormapArg) -> Tuple[str, Colormap]:
             'Must be a {str, tuple, dict, vispy.colormap.Colormap}'
         )
 
-    return name, AVAILABLE_COLORMAPS[name]
+    return AVAILABLE_COLORMAPS[name]
+
+
+def ensure_colormap_tuple(colormap: ValidColormapArg,) -> Tuple[str, Colormap]:
```
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py:428]: will we need this anymore once we have a class that has a `name` attribute?

One thing I was also playing with the other day was mucking with the `__new__` method of the `Colormap` class, such that calling `Colormap(ValidColormapArg)` always does the right thing.  The goal there would be to be able to use `Property[ValidColormapArg, None, Colormap]` as easily as we use our `StringEnums`... thoughts?

File: napari/utils/colormaps/colormap.py (hunk 1 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: we can reproduce all the functionality of all the colormaps we currently have (apart from the two 3D ones that I've never been such a fan of) without `controls` or `interpolation`.

I could see how they might be nice to have, and will be easy to implement, but we could also not

File: napari/utils/colormaps/colormaps.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py:428]: > will we need this anymore once we have a class that has a name attribute?

Nope, just for backwards compatibility right now, i've only converted `Image` but will probably drop this when I'd all done

Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py:428]: > mucking with the __new__ method of the Colormap class, such that calling Colormap(ValidColormapArg) always does the right thing. The goal there would be to be able to use Property[ValidColormapArg, None, Colormap] as easily as we use our StringEnums... thoughts

This sounds quite nice!

Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py:428]: But I will add I think it is also nice to have all the logic for going from string to a colormap outside the colormap object. I.e. `ensure_colormap` is actually quite intense, but `Colormap` right now is very simple. I like them being well separated.

File: napari/utils/colormaps/colormap.py (hunk 1 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: For `controls` the thing I'm not so sure how to do with the dataclass is handle the init where None is provided and then it auto-calculates based on length of `colors` and `interpolation`. In general the dataclasses I find a little more tricky when there is coupling between the attributes.

File: napari/utils/colormaps/colormaps.py (hunk 2 above)
Comment (tlambert03) [reply napari/utils/colormaps/colormaps.py:428]: it could mostly remain that way probably... also, there will probably be some pushback in general to touching `__new__`, since I think people tend not to like that :joy:

File: napari/utils/colormaps/colormap.py (hunk 1 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: I've now added support for controls, interpolation, though i need to test `interpolation='zero'`. Also there are no custom setters now so things could fall out of place once made, but we might want to make this whole thing frozen ...

File: napari/utils/colormaps/colormap.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        Control points of the colormap.
+    interpolation : str
+        Colormap interpolation mode, either 'linear' or
+        'zero'. If 'linear', ncontrols = ncolors (one
+        color per control point). If 'zero', ncontrols
+        = ncolors+1 (one color per bin).
+    """
+
+    name: str = 'undefined'
+    colors: Property[np.ndarray, None, transform_color] = np.zeros((2, 4))
```
Comment (tlambert03) [inline napari/utils/colormaps/colormap.py:38]: suggest either removing colors and controls from the repr with [`field(repr=False)`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field), or declare your own `__repr__`.

File: napari/utils/colormaps/colormap.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+class ColormapInterpolationMode(StringEnum):
+    """INTERPOLATION: Interpolation mode for colormaps."""
+
+    LINEAR = auto()
+    ZERO = auto()
+
+
+@dataclass(events=True, properties=True)
+class Colormap:
```
Comment (tlambert03) [inline napari/utils/colormaps/colormap.py:19]: could be nice to add an `__iter__` method: 
```python
def __iter__(self):
    yield from (self.colors, self.controls, self.interpolation)
```
so that you can easily convert to a vispy colormap with:
```python
cm = napari.utils.Colormap(...)
vispy.colors.Colormap(*cm)
```

Comment (tlambert03) [reply napari/utils/colormaps/colormap.py:19]: btw, it takes 153 µs ± 1.25 µs on my computer to convert to a vispy colormap like that.  no biggie

File: napari/utils/colormaps/colormaps.py [hunk 5]
```diff
@@ -4,13 +4,13 @@
...
     get_colormap,
     get_colormaps,
 )
 
 from ...types import ValidColormapArg
 from .vendored import cm, colorconv
+from .colormap import Colormap
 
 _matplotlib_list_file = os.path.join(
     os.path.dirname(__file__), 'matplotlib_cmaps.txt'
```
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py:16]: in this PR or another one... can we just get rid of `matplotlibs_cmaps.txt`... I get the original inspiration, but it's the only place where we are reading a text file to create a list of strings.

File: napari/utils/colormaps/colormaps.py [hunk 6]
```diff
@@ -273,57 +279,13 @@ def vispy_or_mpl_colormap(name):
                 'or matplotlib.'
             )
         mpl_colors = mpl_cmap(np.linspace(0, 1, 256))
-        cmap = Colormap(mpl_colors)
+        cmap = Colormap(name=name, colors=mpl_colors)
```
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py:282]: perhaps have some `@classmethods` declared on our Colormap (`from_mpl` & `from_vispy`) ... so you can just do, e.g. `cmap = Colormap.from_mpl(mpl_cmap)`

File: napari/utils/colormaps/colormaps.py [hunk 7]
```diff
@@ -393,16 +354,23 @@ def ensure_colormap_tuple(colormap: ValidColormapArg) -> Tuple[str, Colormap]:
...
         name = None
         for key, val in AVAILABLE_COLORMAPS.items():
             if colormap == val:
                 name = key
                 break
         if not name:
             name = increment_unnamed_colormap(AVAILABLE_COLORMAPS)
-        AVAILABLE_COLORMAPS[name] = colormap
+        # Convert from vispy colormap
+        cmap = Colormap(
```
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py:367]: see comment elsewhere about the `Colormap.from_vispy` classmethod idea.

**Review by tlambert03 — COMMENTED** [review]
I'm feeling good about this @sofroniewn, particularly if you like it.  There's only a _tiny_ bit of logic duplication from other libraries, and since they don't all do it the same way, it provides us with a place to unify their APIs internally if need be (see the `@classmethod` comment in review).

It'd be nice to evaluate this without the evented dataclass stuff for now (since we're not really using the events yet).   Since it's a bit hard to use `@property` on a regular python dataclass, maybe just make this a "good old fashioned" class for now (😄 ).  We can pull in the dataclass stuff later.

I'd also like to see the `ensure_colormap_tuple` usage removed.

File: napari/utils/colormaps/colormaps.py (hunk 6 above)
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py:282]: interesting idea, i'll look into that

**Comment by sofroniewn** [comment]
Thanks for the review @tlambert03 glad you're ok with the general direction.

I think I'll make a new PR with a pure class based approach and your additional suggestions, we can then think about the dataclass one later

**Comment by sofroniewn** [comment]
Superceeded by #1523
